patroni/ha.py

```diff
@@ -437,57 +437,65 @@ def is_synchronous_mode(self):
     def is_synchronous_mode_strict(self):
         return self.check_mode('synchronous_mode_strict')
 
-    def process_sync_replication(self):
+    def process_sync_replication(self, sync_node_count=1):
         """Process synchronous standby beahvior.
 
         Synchronous standbys are registered in two places postgresql.conf and DCS. The order of updating them must
-        be right. The invariant that should be kept is that if a node is master and sync_standby is set in DCS,
+        be right. The invariant that should be kept is that if a node is master and sync_standby_list is set in DCS,
         then that node must have synchronous_standby set to that value. Or more simple, first set in postgresql.conf
         and then in DCS. When removing, first remove in DCS, then in postgresql.conf. This is so we only consider
         promoting standbys that were guaranteed to be replicating synchronously.
         """
         if self.is_synchronous_mode():
-            current = self.cluster.sync.leader and self.cluster.sync.sync_standby
-            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster)
-            if picked != current:
-                # We need to revoke privilege from current before replacing it in the config
-                if current:
-                    logger.info("Removing synchronous privilege from %s", current)
-                    if not self.dcs.write_sync_state(self.state_handler.name, None, index=self.cluster.sync.index):
+            current = self.cluster.sync.leader and self.cluster.sync.members or []
+            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster, sync_node_count)
```

Comment:
IMO allow_promote needs rethinking.
With the only possible sync standby it was an indicator that the specific node is already `sync` in pg_stat_replication.
That also means we are [allowed to publish](https://github.com/zalando/patroni/pull/1594/files#diff-57c53e988b775ee2902488fb33c1f433L471-L482) it into the sync key in DCS.

Since the `picked` became the list, the `allow_promote` should be too and it should contain names of nodes from the `picked` that are already synchronous. Only nodes from the allow_promote list are allowed to be published into DCS.

Reply:
It make sense. Will update the code.